### ISSUE TITLE
Show active nav class for "children" of custom menu links

### ISSUE THREAD

**Issue by inostia** [rkhleics/wagtailmenus#215]
Ref #48 

On [this line](https://github.com/rkhleics/wagtailmenus/blob/ddf613244023f9aa1b24b67058d90523084b4e0c/wagtailmenus/models/menus.py#L515) the equality check against path is `==`. Is it possible to change that to `item.link_url in request.path`? Maybe there is a place to add an option so that it isn't forced, but I'm not sure where that would live.

Happy to make a PR if I can get some feedback about this request.

My use-case is that I have some pages that are non-Wagtail/data-driven, including our user directory. We don't want to make Wagtail pages for profiles, but we want to have them under the url `/people/{user}`. I added `/people/` to wagtailmenus and that works fine, but doesn't show active class for the user URL.

**Comment by ababic** [comment]
Hi @inostia,

This is an interesting use case, so thank you for raising. The `'active'` class in wagtailmenus is taken to mean "You are on this page", so the equality test you mentioned above is valid I think, and needs to stay as it is. However, there is another 'active' class in wagtailmenus that is used to indicate that a page is 'an ancestor of the active page', which I think sounds like a better fit for this situation.

The simplest implementation I can think of would be to change the following code:

```
if apply_active_classes and item.link_url == request_path:
    setattr(item, 'active_class', app_settings.ACTIVE_CLASS)
```

To:

```
if apply_active_classes:
    if item.link_url == request_path:
        setattr(item, 'active_class', app_settings.ACTIVE_CLASS)
    elif app_settings.CUSTOM_URL_ANCESTOR_HIGHLIGHTING and request_path.startswith(item.link_url):
        setattr(item, 'active_class', app_settings.ACTIVE_ANCESTOR_CLASS)
```

And add a new setting (`False` by default) to allow developers to turn the behaviour on when needed.

**Comment by inostia** [comment]
Great @ababic that approach makes perfect sense. I did not even consider using the ancestor class. 

I think I'll be able to use some hours at my org this week to get this in, as it's a bit of a blocker for us. I'll reach out with any questions.

**Comment by ababic** [comment]
Thinking about it, it'd probably be more suitable to use `urlparse` to extract the `path` from `item.link_url` to use in both of these checks (which would work even when `item.link_url` included query params or a fragment identifier). We can probably also safely apply active classes for 'relative urls' (where `netloc` is blank).

```
if apply_active_classes:
    if apps_settings.CUSTOM_URL_SMART_ACTIVE_CLASSES:
        # new behaviour
        parsed_url = urlparse(item.link_url)  
        if not parsed_url.netlock:
            if request_path == parsed_url.path:
                setattr(item, 'active_class', app_settings.ACTIVE_CLASS)
            elif request_path.startswith(parsed_url.path):
                setattr(item, 'active_class', app_settings.ACTIVE_ANCESTOR_CLASS)
    elif item.link_url ==  request_path: 
        # previous behaviour
        setattr(item, 'active_class', app_settings.ACTIVE_CLASS)  
```

**Comment by inostia** [comment]
I like the use of `urlparse`. But I can't imagine a scenario in which a person wouldn't want query params/fragment identifiers from being excluded from the match? Also I don't know why it's necessary to check for `netloc` if the user would presumably be on a different domain after clicking the link.

My thinking is for the requested feature, it'd be better to target _specifically_ the case where someone wants to highlight ancestors for custom links, rather than wrapping it under "smart" active classes. That's why I like the setting `CUSTOM_URL_ANCESTOR_HIGHLIGHTING`.

Maybe something like:
```
from urllib.parse import urlparse
...
if apply_active_classes:
    parsed_url = urlparse(item.link_url)
    if app_settings.CUSTOM_URL_ANCESTOR_HIGHLIGHTING and request_path.startswith(parsed_url.path):
        setattr(item, 'active_class', app_settings.ACTIVE_ANCESTOR_CLASS)
    elif parsed_url.path == request_path:
        setattr(item, 'active_class', app_settings.ACTIVE_CLASS)  
```

**Comment by ababic** [comment]
Hi @inostia,

Sorry, yes. I'm attempting to solve a limitation of the current functionality while we're here, and not explaining myself very well. I'm thinking some examples might help.

1. item.link_url is `'/contact-us/#form'` and request_path is `'contact-us/#somewhere-else'`
2. item.link_url is `'/contact-us/#form'` and request_path is `'contact-us/'`
3. item.link_url is `/contact-us/?something=one` and request_path is `'contact-us/?something=two'`
4. item.link_url is `/contact-us/?something=one` and request_path is `'contact-us/'`

If all of these cases, if these were links to pages (with the GET params and fragment identifiers added via `item.url_append`), the `active_class` would always be `app_settings.ACTIVE_CLASS`, because the urls do go to the same page. But currently, none of the above combinations would result in an `app_settings.ACTIVE_CLASS` - hence why I'd like move to `if parsed_url.path == request_path` - I feel like these results should be consistent whether the urls link to pages or custom URLS.

But, this is a change in behaviour - and something that might not be backwards compatible, depending on how wagtailmenus is being used... this is why I introduced the `apps_settings.CUSTOM_URL_SMART_ACTIVE_CLASSES` setting: to allow developers to opt-in to the new functionality if they choose to, but otherwise keep the behaviour the same (by only calling `urlparse` when `apps_settings.CUSTOM_URL_SMART_ACTIVE_CLASSES` is True, I'm also avoiding any performance overhead of calling it when it's not needed - however small that overhead might be).

I added the `if not parsed_url.netloc:` condition originally for the sake of improved efficiency (checking that a string is/isn't blank is faster than doing a string comparison). Since we get this value for free by calling `urlparse` we might as well make use of it, as it would help us avoid unnecessary string comparisons. But, there's actually actually a logical benefit to it, now that I think about it. Take the following cases for example:

5. item.link_url is `'https://website.com/contact-us/'` and request_path is `/contact-us/`
6. item.link_url is `'https://website.com/contact-us/'` and request_path is `/contact-us/offices/`

Without the netloc condition, 5 would result in `app_settings.ACTIVE_CLASS` being applied, and 6 would result in `app_settings.ACTIVE_ANCESTOR_CLASS` - neither of which would I class as desired behaviour.

**Comment by ababic** [comment]
Rather than adjust this code in-place, I would suggest adding a `get_active_class_for_request()` method to `AbstractMenuItem` to handle this particular bit of logic. It'll be much easier to test that way. I would also suggest adding tests for all of the above cases, to be sure it behaves as intended.

**Comment by ababic** [comment]
Hi @inostia. If you'd like a quick fix for this (it'll be approximately 4 weeks until version 2.8 is released), it might be worth experimenting with the [`menus-modify-primed-menu-items`](http://wagtailmenus.readthedocs.io/en/stable/advanced_topics/hooks.html#menus-modify-primed-menu-items) hook in your project.

**Comment by inostia** [comment]
Thanks @ababic . Actually I have a working implementation with tests passing! I'll be making a PR soon.

**Comment by inostia** [comment]
@ababic It's going to be difficult to work out the cases you've mentioned 1 and 2 because the hash fragment does not get sent to the server, and so is unavailable in the `HttpRequest` object. We can get the query params for 3 and 4 and compare those, but not the hashes as far as I can tell.


I'll make the PR as is and if there's any way you can think to compare hashes I'm happy to look deeper into it. I'm hesitant to implement the query params check as it would be inconsistent with handling of hash fragments.

**Comment by ababic** [comment]
Hi @inostia. Thanks for your efforts so far. I think what we're really looking to test is that in each case, the method always takes the `path` from the supplied request (which will always be void of query params or fragments) and compares it to `parsed_url.path` to get a result. It sounds like there's not much value in testing both 1 and 2, but I think 2 is worthwhile. The value in testing both 3 and 4 is perhaps questionable too,  but simulating at least one request with query params I think would be worthwhile.

**Comment by inostia** [comment]
Thanks @ababic. You can find the relevant tests [here](https://github.com/rkhleics/wagtailmenus/pull/221/commits/71f5bfc525b22f58aebe02a962272ffaae896788). I think this is what you mean. Please look over the PR.

**Comment by inostia** [comment]
Fixed in #221
